### DISCUSSION TITLE
javax.xml.transform.TransformerException

### DISCUSSION THREAD

**Discussion by TedDawkin** [freeplane/freeplane#295] [Help] [answered]
**Answer: With the current version 1.9.14 there is no way to prevent the exception. Workraound is to write a groovy-script and clean the html as a form of HtmlSanitizer**

Hi,

when I try to export my mindmap to html with java or any java-related export (also export to .odt) , I get this exception:

**javax.xml.transform.TransformerException: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: The element type "div" must be terminated by the matching end-tag "</div>".**

I wished Freeplane would have an option to just ignore this exception when exporting.

Reason for Exception: I have a lot of nodes where I used "Edit node in Dialog" and did a copy and paste from websites - so insede the Text of the edit node Dialog is HTML with missing tags here and there.

Reason for Export:
I need to export the mindmap in Java or a clickable version because I want to transfer my project to a notebook where no installation of any software is allowed.

An export to plain html or pdf or jpeg is not enough because I want to navigate my mindmap and read the text I added to the nodes. Plain PDF or HTML is to confusing because it is a  lot of thet and you can not find what you need with no nodes and linked nodes that are expandable

How can I export the mindmap?

I tried parsing the *.mm via different Freeplane\resources\xslt\mm2---.xsl and parsing/finding the missing end-tags but it is to much data and to much effort to solve this within a couple of hours - I found some concept of "Sanitize Html" or HtmlSanatizer to ignore ar add missing tags but do not know how to do this with an mindmap or xml

Example:
[example2.zip](https://github.com/freeplane/freeplane/files/8420728/example2.zip)
.

**Comment by EdoFro** [comment]
> Reason for Export:
I need to export the mindmap in Java or a clickable version because I want to transfer my project to a notebook where no installation of any software is allowed.

You can install [PortableApps ](https://portableapps.com/) and install the portable version of Freeplane (.paf.exe) in it. You can do this in a pendrive and run freeplane from there.

**Reply by TedDawkin** [reply]
Thank you for the idea - I forgot to mention this: any type of .exe will not run on this notebook that is not whitelisted - so portable apps do not work. :(   

why is this notebook so important? it is allowed to be used in an exam at university. you can have all sorts of pdf, doc, html to use in the final exam - it was planed to write this exam at home [because corona] so all my notes are in freeplane an in the last exam (that we wrote at home at our own pc) it was a bliss to have it.

 Now it was decided that our study-notebook is used

**Reply by EdoFro** [reply]
I tried with different export options and I had the error too. But with the HTML option worked ok:

![image](https://user-images.githubusercontent.com/35700575/161843651-1459bb77-5aa7-4c50-9b2f-b28a44fde18b.png)

![image](https://user-images.githubusercontent.com/35700575/161843864-f0cbd17f-5d52-48fa-80cc-784738e0a01b.png)

and it opens in chrome

**Reply by TedDawkin** [reply]
Hi EdoFro,

as I wrote in the thread-opener:

An export to plain html or pdf or jpeg is not enough [insufficient] because I want to navigate my mindmap and read the text I added to the nodes. Plain PDF or HTML is to confusing because it is a lot of thet and you can not find what you need with no nodes and linked nodes that are expandable - the mindmap is to complex and with plain html/pdf/odf I am not able to find stuff in time.

Thank you for investing your work - I should have made that clearer in the opener.

**Reply by EdoFro** [reply]
Hi TedDawkin,
In this HTML format, the map get exported with buttons to expand and fold the nodes. I thought that could be useful for you.
bye,

**Comment by rexelbartolome** [comment]
that's tough, the last thing on my mind is publishing the map online but both options requires an executable to work (e.g., Java Applet and base Freeplane)
https://www.freeplane.org/wiki/index.php/Publish_mindmap_on_the_internet

Seems a bit of a stretch but you could somehow convince the administrators that the program you're using is open source and therefore safe? 

You could also look into exporting the mind map and importing it in other online mindmapping websites but they'll lose a bit of functionality compared to using native freeplane, which could be a deal breaker since you mentioned expandable nodes and links etc... Don't know how compatible freeplane's export is to their import either. So it's likely you'll redo a significant part of the mindmap in your website of choice

**Comment by TedDawkin** [comment]
I do not see a Workaround as the right approach - also: Internet will not be available.

Honestly if I am looking for a result X I am not looking for a Workaround Y - or said in other words: If I am in need to make a pizza, I do not want to see Pasta.

I was more hoping for a way to supress the exception - because a missing tag "is no problem" - the html is displayed correctly in Freeplane and exporter correctly as simple HTML even with a missing tag, so why not "ignore" this exception. 

Thank you for your ideas.

**Comment by EdoFro** [comment, marked as answer]
ok. this is not _**Result X**_ but it may help:

```groovy
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.safety.Whitelist


def nodos=node.map.root.find{ n->
    n.note?true:false
}

nodos.each{ n ->
//def n=nodos[0]
    Document doc = Jsoup.parse(n.noteText);
    n.noteText = null
    def myWhitelist =  Whitelist.relaxed()
    def cleaned = Jsoup.clean(doc.body().html(), myWhitelist)
    n.noteText = Jsoup.parse(cleaned).html()
}
```

It's like a pizza Hawaii or a Calzonne: not really a pizza but many of us are happy to get that served (specially when it's for free)

**Reply by EdoFro** [reply]
PD:
- it worked with your example file.
- after running the script I was able to export the file as XML
- I attach the modiifed .mm and XML file

[example2.zip](https://github.com/freeplane/freeplane/files/8444233/example2.zip)

**Comment by TedDawkin** [comment]
I would like to try this. At the moment I get an error.

I created a file in a text editor. [Saved it into the Freeplane Scripts directory and tried to run it inside freeplane...](https://www.freeplane.org/wiki/index.php/Scripting) 

Did I do it right? I guess I am missing something with the import

![grafik](https://user-images.githubusercontent.com/21216968/162238991-ad5dc5d8-8553-4ed6-89a3-3167a8ffef62.png)

![grafik](https://user-images.githubusercontent.com/21216968/162241127-ee822364-2616-4c77-9bf4-c5e1d95660b3.png)

I created another test-schript
with node.text = "Hello World!" and this worked

If i try

`def nodos=node.map.root.find{ n->
    n.note?true:false
}`

No error until I add the **import** lines

**Reply by EdoFro** [reply]
I have no time right now. Can you see if in the log.0 file was logged an error? Can you attach it?

**Reply by EdoFro** [reply]
maybe you have to add the jsoup library to freeplane

**Comment by TedDawkin** [comment]
I have taken enough of your time. Thank you for your help.
I forgot to look in the log. It says " unable to resolve class org.jsoup.Jsoup  @ line 1, column 1.    import org.jsoup.Jsoup"
I will try to find a way to add the library and post/mark a solution.

Knowing that a script is the right way helped me!

**Solutio to import**

- [Download missing jsoup-jar](https://jsoup.org/packages/jsoup-1.14.3.jar)
- Copy jsoup-1.14.3.jar to AppData\Roaming\Freeplane\1.9.x\lib
- Install [scriptlib-addon](https://www.freeplane.org/wiki/index.php/Add-ons_(install)#scriptlib)
- In Tools >Preferences > Plugins classpath must be set to: lib with read/write rights checked
- Restart freeplane and run sanitizing script
- Last step: Praise EdoFro :)

**Reply by EdoFro** [reply]
Glad it helped. ;)

Here is another, maybe easier, way (I don't know if it works because I have the library allready installed, and I don't know how to test it)

```groovy
// https://mvnrepository.com/artifact/org.jsoup/jsoup
@Grab(group='org.jsoup', module='jsoup', version='1.14.3')

import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.safety.Safelist


def nodos=node.map.root.find{ n->
    n.note?true:false
}

nodos.each{ n ->
//def n=nodos[0]
    Document doc = Jsoup.parse(n.noteText);
    n.noteText = null
    def myWhitelist =  Safelist.relaxed()
    def cleaned = Jsoup.clean(doc.body().html(), myWhitelist)
    n.noteText = Jsoup.parse(cleaned).html()
}
```

As far as I know, the `@Grab`  command downloads and installs the library for you, the first time you run the script.
